larpix/larpix.py

```diff
@@ -11,38 +11,214 @@
 import math
 import warnings
 import struct
+import sys
 
 from bitarray import bitarray
 
 from . import bitarrayhelper as bah
 from .logger import Logger
 from . import configs
 
+class Key(object):
+    '''
+    A unique specification for routing data to a particular detector sub-system.
+    At the core, a key is represented by 3-unsigned 1-byte integer fields which
+    refer to an id code within a layer of the LArPix DAQ system heirarchy.
+    Field 0 represents the io group id number, field 1 represents the io
+    channel connecting to a MISO/MOSI pair, and field 2 represents the chip
+    id. The io group is the device controlling a set of MOSI/MISO pairs, the
+    io channel is a single MOSI/MISO pair controlling a collection of LArPix
+    asics, and the chip id uniquely identifies a chip on a single MISO/MISO
+    network.
+
+    Each field should be a 1-byte unsigned integer (0-255) providing a unique
+    lookup value for each component in the system. The id values of 0 and 255
+    are reserved for special functionality.
+
+    A key can be specified by a string of ``'<io group>-<io channel>-<chip id>'`` or by
+    using other Keys.
+
+    Keys are hashed by their string representation and are equivalent to their
+    string representation so::
+
+        key = Key('1-1-1')
+
+        key == '1-1-1' # True
+
+        d = { key: 'example' }
+        d[key] == 'example' # True
+        d['1-1-1'] == 'example' # True
+
+    '''
+    key_delimiter = '-'
+    key_format = key_delimiter.join(('{io_group}', '{io_channel}', '{chip_id}'))
+
+    def __init__(self, key):
+        if isinstance(key, bytes):
+            self.keystring = str(key.decode("utf-8"))
+        else:
+            self.keystring = str(key)
+
+    def __repr__(self):
+        return 'Key(keystring=\'{}\')'.format(self.keystring)
```

Comment:
``repr`` should return a string that could be executed, but ``Key(keystring='1-2-3')`` will not execute because the parameter is named ``key`` in ``__init__``. I'd recommend just ``Key('1-2-3')``